metaserver/src/main/scala/scala/meta/languageserver/search/SqliteSymbolIndex.scala

```diff
@@ -0,0 +1,178 @@
+import scala.meta.languageserver.index.SymbolData
+import com.typesafe.scalalogging.LazyLogging
+import langserver.core.Notifications
+import langserver.{types => l}
+import org.langmeta.internal.semanticdb.{schema => s}
+import org.langmeta.io.AbsolutePath
+import org.langmeta.semanticdb.Symbol
+
+class SqliteSymbolIndex(
+    cwd: AbsolutePath,
+    notifications: Notifications,
+    buffers: Buffers,
+    serverConfig: ServerConfig,
+) extends SymbolIndex with LazyLogging {
+  private val conn: Option[Connection] = {
+    val connPath = cwd.resolve(".metaserver").resolve("semanticdb.sqlite")
+    val connString = s"jdbc:sqlite:$connPath"
+    try {
+      val conn = DriverManager.getConnection(connString)
+      logger.info(s"Successfully initialized connection to $connString")
+      Some(conn)
+    } catch {
+      case ex: Throwable =>
+        logger.error(s"Failed to initialize connection to $connString", ex)
+        None
+    }
+  }
+
+  private val findSymbolSql =
+    """|select s.symbol
+       |from document as d, name as n, symbol as s
+       |where d.filename == ?
+       |and n.document == d.id
+       |and n.start_line <= ?
+       |and n.start_character <= ?
+       |and n.end_line >= ?
+       |and n.end_character >= ?
+       |and s.id=n.symbol""".stripMargin
+  private val findSymbolStmt = conn.map(_.prepareStatement(findSymbolSql))
+
+  def findSymbol(path: AbsolutePath, line: Int, column: Int): Option[Symbol] = {
+    findSymbolStmt match {
+      case Some(findSymbolStmt) =>
+        findSymbolStmt.setString(1, path.toRelative(cwd).toString)
+        findSymbolStmt.setInt(2, line)
+        findSymbolStmt.setInt(3, column)
+        findSymbolStmt.setInt(4, line)
+        findSymbolStmt.setInt(5, column)
+        val findSymbolRs = findSymbolStmt.executeQuery()
+        try {
+          while (findSymbolRs.next()) {
+            val s_symbol = findSymbolRs.getString(1)
+            return Some(Symbol(s_symbol))
+          }
+          None
+        } finally {
+          findSymbolRs.close()
+        }
+      case _ =>
+        None
+    }
+  }
+
+  private val symbolIdSql = "select s.id from symbol as s where s.symbol == ?"
+  private val definitionDataSql =
+    """|select d.filename, n.start_line, n.start_character, n.end_line, n.end_character
+       |from name as n, document as d
+       |where n.symbol == ?
+       |and n.is_definition == 1
+       |and d.id == n.document""".stripMargin
+  private val symbolIdStmt = conn.map(_.prepareStatement(symbolIdSql))
+  private val definitionDataStmt = conn.map(_.prepareStatement(definitionDataSql))
+
+  def definitionData(symbol: Symbol): Option[SymbolData] = {
+    (symbolIdStmt, definitionDataStmt) match {
+      case (Some(symbolIdStmt), Some(definitionDataStmt)) =>
+        // TODO: Take into account symbol.definitionAlternative.
+        symbolIdStmt.setString(1, symbol.toString)
+        val symbolIdRs = symbolIdStmt.executeQuery()
```

Comment:
I can vouch for quill: extremely pleasant to work with! That said, I would make sure it works reliably with sqlite, which is often a bit "strange" compared to other dbs. I've had problems with slick and sqlite too in the past.

Reply:
One more vote for quill, it has been my favorite db lib to use so far.

Reply:
@fwbrasil Can you comment on Quill's compatibility with SQLite?

Reply:
Quill has the dialect implementation for SQLite but I've seen a complaint on Twitter that it has bugs. I guess not many people use Quill with SQLite since there are no open bugs for it. I can review the dialect later this week if you're planning to use it.

Another alternative is H2. It has better performance, can be used in a client-server setup if necessary, and at least supports standard SQL.

Reply:
I have no personal attachment nor preference for sqlite over h2. It sounds like we should consider using h2, could you check how h2 impacts performance/space @xeno-by ?

@fwbrasil Perhaps unrelated, but maybe you have some insights on a problem we're facing: how do we compose indices from different sources? I imagine this could impact the choice of a db, but I'm quite a noob in this space 😅. For example, we want to be able to query for all references of a symbol across library1.jar:library2.jar:test-classes/

- We could store everything in a single database per user, but that would mean concurrent writes since indexing independent libraries can be done in parallel.
- We could store everything in a single database per workspace, but that could mean a lot of wasted disk space for machines with many projects. I personally have 587 projects in my `~/dev` directory, so this is a no-go for me.
- We could separate each index by its source (`library1.jar.db`), and for each configuration we open connections to each database per entry in the classpath. I'm not sure how practical this approach is for large classpaths or evolving projects with moving modules.

I'm personally leaning towards option 1 if we can take advantage of concurrent writes. I had previously outruled this option because sqlite doesn't seem to support concurrent writes.

Reply:
Both H2 and HSQLDB embedded dbs support multiple writers using mvcc.

Reply:
I didn't know about sqlite's WAL, nice find! 

Reading up a bit, it seems we should also consider performance of deletions. We will need to regularly clean up semanticdb indices from stale snapshots of source files. It would be good to investigate difference between h2/sqlite with regards to deletions @xeno-by